src/binlog/events/transaction_payload_event.rs

```diff
@@ -0,0 +1,296 @@
+// modified, or distributed except according to those terms.
+
+use crate::io::ReadMysqlExt;
+use std::{borrow::Cow, cmp::min, convert::TryFrom, io};
+
+use saturating::Saturating as S;
+
+use super::BinlogEventHeader;
+use crate::{
+    binlog::{
+        consts::{
+            BinlogVersion, EventType, TransactionPayloadCompressionType, TransactionPayloadFields,
+        },
+        BinlogCtx, BinlogEvent, BinlogStruct,
+    },
+    io::ParseBuf,
+    misc::raw::{bytes::EofBytes, int::*, RawBytes},
+    proto::{MyDeserialize, MySerialize},
+};
+
+/// The rotate event is added to the binlog as last event
+/// to tell the reader what binlog to request next.
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct TransactionPayloadEvent<'a> {
+    // payload size
+    payload_size: RawInt<LeU64>,
+
+    // compression algorithm
+    algorithm: TransactionPayloadCompressionType,
+
+    // uncompressed size
+    uncompressed_size: RawInt<LeU64>,
+
+    // payload to be decompressed
+    payload: RawBytes<'a, EofBytes>,
+
+    // size of parsed header
+    header_size: usize,
+}
+
+impl<'a> TransactionPayloadEvent<'a> {
+    pub fn new(
+        payload_size: u64,
+        algorithm: TransactionPayloadCompressionType,
+        uncompressed_size: u64,
+        payload: impl Into<Cow<'a, [u8]>>,
+    ) -> Self {
+        Self {
+            payload_size: RawInt::new(payload_size),
+            algorithm: algorithm,
+            uncompressed_size: RawInt::new(uncompressed_size),
+            payload: RawBytes::new(payload),
+            header_size: 0,
+        }
+    }
+
+    /// Sets the `payload_size` field value.
+    pub fn with_payload_size(mut self, payload_size: u64) -> Self {
+        self.payload_size = RawInt::new(payload_size);
+        self
+    }
+    /// Sets the `algorithm` field value.
+    pub fn with_algorithm(mut self, algorithm: TransactionPayloadCompressionType) -> Self {
+        self.algorithm = algorithm;
+        self
+    }
+    /// Sets the `uncompressed_size` field value.
+    pub fn with_uncompressed_size(mut self, uncompressed_size: u64) -> Self {
+        self.uncompressed_size = RawInt::new(uncompressed_size);
+        self
+    }
+
+    /// Sets the `payload` field value.
+    pub fn with_payload(mut self, payload: impl Into<Cow<'a, [u8]>>) -> Self {
+        self.payload = RawBytes::new(payload);
+        self
+    }
+
+    /// Returns the payload_size.
+    pub fn payload_size(&self) -> u64 {
+        self.payload_size.0
+    }
+
+    /// Returns raw payload of the binlog event.
+    pub fn payload_raw(&'a self) -> &'a [u8] {
+        self.payload.as_bytes()
+    }
+
+    /// Returns raw payload decompressed
+    pub fn payload_raw_decompressed(self) -> Vec<u8> {
+        if self.algorithm == TransactionPayloadCompressionType::NONE {
+            return self.payload_raw().to_vec();
+        }
+        let decoded = zstd::stream::decode_all(self.payload.as_bytes()).unwrap();
```

Comment:
Can we use `decompressed_size` to preallocate a buffer for `zstd::stream::copy_decode` here?

Reply:
Done